File: backend/src/models/Etablissement.js
const mongoose = require('mongoose');
const { ETAB_STATUS, ETAB_TYPES } = require('../utils/constants');

const etablissementSchema = new mongoose.Schema({
  nom: { type: String, required: true, trim: true },
  type: { type: String, enum: ETAB_TYPES, required: true },
  description: { type: String, default: '' },
  adresse: {
    rue: { type: String, required: true },
    ville: { type: String, required: true },
    codePostal: String,
    coordonnees: {
      type: { type: String, enum: ['Point'], default: 'Point' },
      coordinates: { type: [Number], default: [0, 0] },
    },
  },
  telephone: { type: String, required: true },
  email: String,
  photo: { type: String, default: null },
  responsable: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
  documents: [{ nom: String, url: String, uploadedAt: Date }],
  statut: { type: String, enum: Object.values(ETAB_STATUS), default: ETAB_STATUS.PENDING },

  horaires: {
    lundi:    { ouvert: { type: Boolean, default: false }, debut: String, fin: String },
    mardi:    { ouvert: { type: Boolean, default: false }, debut: String, fin: String },
    mercredi: { ouvert: { type: Boolean, default: false }, debut: String, fin: String },
    jeudi:    { ouvert: { type: Boolean, default: false }, debut: String, fin: String },
    vendredi: { ouvert: { type: Boolean, default: false }, debut: String, fin: String },
    samedi:   { ouvert: { type: Boolean, default: false }, debut: String, fin: String },
    dimanche: { ouvert: { type: Boolean, default: false }, debut: String, fin: String },
  },

  fermetures: [{ date: Date, motif: String }],
  noteMoyenne: { type: Number, default: 0 },
  nombreAvis: { type: Number, default: 0 },

  abonnement: {
    type: { type: String, enum: ['gratuit', 'standard', 'premium'], default: 'gratuit' },
    expiresAt: Date,
  },
}, { timestamps: true });

etablissementSchema.index({ 'adresse.coordonnees': '2dsphere' });
etablissementSchema.index({ nom: 'text', description: 'text' });

module.exports = mongoose.model('Etablissement', etablissementSchema);

File: mobile-app/lib/models/etablissement.dart
class Etablissement {
  final String id, nom, type, description, telephone, statut;
  final String? email, photo;
  final Adresse adresse;
  final double noteMoyenne;
  final int nombreAvis;
  final Map<String, Horaire> horaires;

  Etablissement({required this.id, required this.nom, required this.type, required this.description,
    required this.telephone, required this.statut, this.email, this.photo,
    required this.adresse, required this.noteMoyenne, required this.nombreAvis, required this.horaires});

  factory Etablissement.fromJson(Map<String, dynamic> json) {
    Map<String, Horaire> h = {};
    if (json['horaires'] != null) {
      (json['horaires'] as Map<String, dynamic>).forEach((k, v) {
        h[k] = Horaire.fromJson(v);
      });
    }
    return Etablissement(
      id: json['_id'] ?? '', nom: json['nom'] ?? '', type: json['type'] ?? '',
      description: json['description'] ?? '', telephone: json['telephone'] ?? '',
      statut: json['statut'] ?? '', email: json['email'], photo: json['photo'],
      adresse: Adresse.fromJson(json['adresse'] ?? {}),
      noteMoyenne: (json['noteMoyenne'] ?? 0).toDouble(),
      nombreAvis: json['nombreAvis'] ?? 0, horaires: h,
    );
  }
}

class Adresse {
  final String rue, ville;
  final double lat, lng;
  Adresse({required this.rue, required this.ville, required this.lat, required this.lng});
  factory Adresse.fromJson(Map<String, dynamic> json) {
    final coords = json['coordonnees']?['coordinates'] ?? [0, 0];
    return Adresse(rue: json['rue'] ?? '', ville: json['ville'] ?? '', lng: (coords[0] ?? 0).toDouble(), lat: (coords[1] ?? 0).toDouble());
  }
}

class Horaire {
  final bool ouvert;
  final String? debut, fin;
  Horaire({required this.ouvert, this.debut, this.fin});
  factory Horaire.fromJson(Map<String, dynamic> json) => Horaire(ouvert: json['ouvert'] ?? false, debut: json['debut'], fin: json['fin']);
}
